build.assets/tooling/cmd/difftest/main.go

```diff
@@ -0,0 +1,210 @@
+)
+
+func main() {
+	kingpin.Version("0.0.1")
+	command := kingpin.Parse()
+
+	// Set default git directory to cwd
+	if repoPath == nil {
+		p, err := os.Getwd()
+		if err != nil {
+			bail(trace.Errorf("Error getting current working directory: %v", err))
+		}
+
+		repoPath = &p
+	}
+
+	// Check if git is available
+	err := gitIsAvailable()
+	if err != nil {
+		bail(trace.Wrap(err, "git is not available"))
+	}
+
+	start := time.Now()
+
+	// Get fork commit ref
+	ref, err := gitMergeBase(*repoPath, *branch)
+	if err != nil {
+		bail(trace.Wrap(err, "fork point might not exist"))
+	}
+
+	// Get git diff with fork commit
+	changes, err := gitChanges(*repoPath, ref)
+	if err != nil {
+		bail(err)
+	}
+
+	// Get a list of changed files
+	changedFiles, err := getChangedTestFilesFromDiff(changes, *exclude, *include)
+	if err != nil {
+		bail(err)
+	}
+
+	switch command {
+	case "diff":
+		diff(*repoPath, ref, changedFiles, time.Since(start))
+	case "test":
+		test(*repoPath, ref, changedFiles)
+	}
+}
+
+// diff prints diff for debug purposes
+func diff(repoPath string, ref string, changedFiles []string, elapsed time.Duration) {
+	fmt.Printf("Tests changed in %v compared to %v:\n\n", repoPath, ref)
+
+	if len(changedFiles) == 0 {
+		fmt.Println("No changes!")
+		return
+	}
+
+	inspect(repoPath, ref, changedFiles, func(filename string, r CompareResult) {
+		if !r.HasNew() && !r.HasChanged() {
+			return
+		}
+
+		fmt.Printf("- %v:\n", filename)
+
+		for _, n := range r.New {
+			fmt.Printf("  +%v (%v)\n", n.Name, n.RefName)
+		}
+
+		for _, n := range r.Changed {
+			fmt.Printf("  ~%v (%v)\n", n.Name, n.RefName)
+		}
+
+		fmt.Println()
+	})
+
+	fmt.Printf("Time elapsed: %s\n", elapsed)
+}
+
+// test builds and prints go test flags
+func test(repoPath string, ref string, changedFiles []string) {
+	var dirs = make(StringSet)
+	methods := make([]string, 0)
+
+	dollar := "$"
+	if *doubleDollar {
+		dollar = "$$"
+	}
+
+	inspect(repoPath, ref, changedFiles, func(filename string, r CompareResult) {
+		if !r.HasNew() && !r.HasChanged() {
+			return
+		}
+
+		for _, n := range r.New {
+			methods = append(methods, "^"+n.RefName+dollar)
+			dirs[relDirName(filename, *relative)] = struct{}{}
+		}
+
+		if *excludeUpdates {
+			return
+		}
+
+		for _, n := range r.Changed {
+			methods = append(methods, "^"+n.RefName+dollar)
+			dirs[relDirName(filename, *relative)] = struct{}{}
+		}
+	})
+
+	if len(methods) == 0 || len(dirs) == 0 {
+		os.Exit(0)
+	}
+
+	fmt.Printf(`-run "%v"`, strings.Join(methods, "|"))
+	if !*onlyRunFlag {
+		fmt.Printf(" %v", strings.Join(dirs.Keys(), " "))
+	}
+}
+
+// inspect iterates over changes in the repo
+func inspect(repoPath string, ref string, changedFiles []string, fn func(string, CompareResult)) {
+	runners, err := findAllSuiteRunners(repoPath, changedFiles)
+	if err != nil {
+		bail(err)
+	}
+
+	for _, filename := range changedFiles {
+		file, err := gitGetFileFromRevision(repoPath, filename, ref)
+		if err != nil {
+			bail(err)
```

Comment:
It would be nice to include the name of the file where we encountered an error in the error message, this way if we run into corner cases where the tool fails we can investigate what went wrong.